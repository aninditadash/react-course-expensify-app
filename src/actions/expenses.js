import { v4 as uuidv4 } from "uuid";

import database from "../firebase/firebase";
import { ref, set, push, get, remove, update } from "firebase/database";

// Action generators for EXPENSES

// Normal application flow in redux
// component calls action generator
// component dispatches object
// redux store changes
//

// saving data to Firebase ----->
// component calls action generator
// component dispatches function
// function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_EXPENSE
// export const addExpense = ({
//   description = "",
//   note = "",
//   amount = 0,
//   createdAt = 0
// } = {}) => ({
//   type: "ADD_EXPENSE",
//   expense: {
//     id: uuidv4(),
//     description,
//     note,
//     amount,
//     createdAt
//   }
// });

// ADD_EXPENSE
export const addExpense = (expense = {}) => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAppExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };
    const dataRef = push(ref(database, "expenses"));
    return set(dataRef, expense).then(() => {
      dispatch(
        addExpense({
          id: dataRef.key,
          ...expense
        })
      );
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id = "" } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id = "" }) => {
  return (dispatch) => {
    return remove(ref(database, `expenses/${id}`)).then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return update(ref(database, `expenses/${id}`), updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return (dispatch) => {
    const dataRef = ref(database, "expenses");
    return get(dataRef)
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      })
      .catch((e) => console.log("Error fetching data:: ", e));
  };
};
