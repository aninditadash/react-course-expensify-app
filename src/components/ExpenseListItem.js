import React from "react";
import moment from "moment";
import numeral from "numeral";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <Link className="expense-list-item" to={`/edit/${id}`}>
    <div>
      <p className="expense-list-item__title expense-list-item--bold">
        {description}
      </p>
      <p className="expense-list-item__subtitle">
        {moment(createdAt).format("MMMM Do, YYYY")}
      </p>
    </div>
    <p className="expense-list-item__data expense-list-item--bold">
      {numeral(amount).format("$0,0.00")}
    </p>
  </Link>
);

export default ExpenseListItem;
