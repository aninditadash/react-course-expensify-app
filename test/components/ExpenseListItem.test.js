import React from "react";
import { shallow } from "enzyme";
import ExpenseListItem from "../../src/components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("ExpenseListItem Component: Should render ExpenseListItem with expense 1", () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test("ExpenseListItem Component: Should render ExpenseListItem with expense 2", () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});
