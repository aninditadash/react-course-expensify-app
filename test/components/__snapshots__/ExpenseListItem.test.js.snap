// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItem Component: Should render ExpenseListItem with expense 1 1`] = `
<Link
  className="expense-list-item"
  to="/edit/1"
>
  <div>
    <p
      className="expense-list-item__title expense-list-item--bold"
    >
      Water bill
    </p>
    <p
      className="expense-list-item__subtitle"
    >
      January 1st, 1970
    </p>
  </div>
  <p
    className="expense-list-item__data expense-list-item--bold"
  >
    $4,500.00
  </p>
</Link>
`;

exports[`ExpenseListItem Component: Should render ExpenseListItem with expense 2 1`] = `
<Link
  className="expense-list-item"
  to="/edit/2"
>
  <div>
    <p
      className="expense-list-item__title expense-list-item--bold"
    >
      Rent
    </p>
    <p
      className="expense-list-item__subtitle"
    >
      December 28th, 1969
    </p>
  </div>
  <p
    className="expense-list-item__data expense-list-item--bold"
  >
    $2,500.00
  </p>
</Link>
`;
